rust-runtime/aws-smithy-wasm/src/wasi.rs

```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! HTTP WASI Adapter
+use aws_smithy_http::header::ParseError;
+use aws_smithy_runtime_api::{
+    client::{
+        http::{
+            HttpClient, HttpConnector, HttpConnectorFuture, HttpConnectorSettings,
+            SharedHttpClient, SharedHttpConnector,
+        },
+        orchestrator::HttpRequest,
+        result::ConnectorError,
+        runtime_components::RuntimeComponents,
+    },
+    http::Response,
+    shared::IntoShared,
+};
+use aws_smithy_types::body::SdkBody;
+use bytes::{Bytes, BytesMut};
+use wasi::http::{
+    outgoing_handler,
+    types::{self as wasi_http, OutgoingBody, RequestOptions},
+};
+
+/// Creates a connector function that can be used during instantiation of the client SDK
+/// in order to route the HTTP requests through the WebAssembly host. The host must
+/// support the WASI HTTP proposal as defined in the Preview 2 specification.
+pub fn wasi_http_client() -> SharedHttpClient {
+    WasiHttpClient::new().into_shared()
+}
+
+/// HTTP client used in WASI environment
+#[derive(Debug, Clone)]
+pub struct WasiHttpClient {
+    connector: SharedHttpConnector,
+}
+
+impl WasiHttpClient {
+    /// Create a new Wasi HTTP client.
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+impl Default for WasiHttpClient {
+    fn default() -> Self {
+        Self {
+            connector: WasiHttpConnector.into_shared(),
+        }
+    }
+}
+
+impl HttpClient for WasiHttpClient {
+    fn http_connector(
+        &self,
+        _settings: &HttpConnectorSettings,
+        _components: &RuntimeComponents,
+    ) -> SharedHttpConnector {
+        self.connector.clone()
+    }
+}
+
+/// HTTP connector used in WASI environment
+#[non_exhaustive]
+#[derive(Debug)]
+pub struct WasiHttpConnector;
+
+impl HttpConnector for WasiHttpConnector {
+    fn call(&self, request: HttpRequest) -> HttpConnectorFuture {
+        tracing::trace!("WasiHttpConnector: sending request {request:?}");
+        // TODO(wasi): add connect/read timeouts
```

Comment:
Is this going to be a separate PR, and we'll just say don't use the WASI client in production yet?

Reply:
Discussed offline and I ended up taking the timeouts from the `HttpConnectorSettings` and baking them into the client, so this is taken care of